### DISCUSSION TITLE
Can I custom the transaction error message?

### DISCUSSION THREAD

**Discussion by knota-lab** [diesel-rs/diesel#2859] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/26835087/129743276-f46c858e-3cc6-4460-a325-adbe36c677e5.png)

**Comment by weiznich** [comment, marked as answer]
Please include some more details about what you actually want to know instead of an image.

**Reply by knota-lab** [reply]
I want return a custom error in transaction. Such as `return Err("Invalid version")`.
But as far as I know it only support `diesel::result::Error`.
Is there any way to return a custom error message in transaction?

**Reply by weiznich** [reply]
As you can see by looking at the function signature of [`Connection::transaction`](https://docs.diesel.rs/1.4.x/diesel/connection/trait.Connection.html#method.transaction) it does support any error type as long as it implements `From<diesel::result::Error>`. You just need to specify which type should be used as part of the generic arguments of that function or write your code in such a way that rustc is able to infer exactly one type.

**Reply by knota-lab** [reply]
Thanks a lot.

```rust
#[derive(Debug)]
struct CusErr {
    pub message: String
}

impl From<Error> for CusErr {
    fn from(_: Error) -> Self {
        CusErr {
            message: "Invalid version".to_string()
        }
    }
}

if old_doc.object_version_number != value.object_version_number {
    return Err(CusErr::from(Error::NotFound));
}
```
